### PR TITLE
Refactor(counters): Improve tests, docs and readability

### PR DESCRIPTION
**Pull request by SandeepTuniki**
(Part of a series of general improvements I am doing with the help of Gemini CLI. These changes do not introduce any new functionality but make the existing code easier to understand, use, and maintain.)

This PR improves the `util/counters.py` module with refactoring, improved documentation, and more comprehensive testing.

### Key Changes

- **Refactoring:**
    - Refactored the `Counters` class and its tests for better readability and focus.
- **Documentation:**
    - Added detailed docstrings with usage examples to all methods in `Counters` class.
- **Testing:**
    - Improved test coverage by adding more granular tests for existing functionality and new tests for edge cases and previously untested methods (`add_counters`, `set_prefix`, rate-limiting logic, etc.).

### CODE REVIEW

File: util/counters.py [hunk 1]
```diff
@@ -39,25 +47,29 @@ class CounterOptions(NamedTuple):
...
+    This class provides a flexible way to handle various types of counters,
+    including min/max values, debug counters, and periodic metrics like
+    processing rate, memory usage, and CPU time.
 
     Example usage:
       counters = Counters(prefix='my_process')
       ...
       counters.add_counter('input_rows')
-         .add_counter('output_rows', 10)
+      counters.add_counter('output_rows', 10)
```
Comment (ajaits) [inline util/counters.py:60]: this was to illustrate chaining allowed

**Review by ajaits — APPROVED** [review]

File: util/counters.py (hunk 1 above)
Comment (SandeepTuniki) [reply util/counters.py]: reverted this change to the chaining calls.
